[0-yibai] iP

0-yibai opened the pull request:
### Wis

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Wis frees your mind of having to remember things you need to do. It's,
- **text-based**
- easy to learn
- ~~convenient to use~~

All you need to do is:

1. download it from [here](https://github.com/0-yibai/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😊

And it's **FREE**!

Features:

- [ ] Managing _todos_
- [ ] Managing _deadlines_
- [ ] Managing _events_
- [ ] Marking and unmarking tasks as done
- [ ] Searching for tasks using keywords using `find <keyword>`
- [ ] Deleting tasks using `delete <task_number>`
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

0-yibai committed:
gave the chatbot a new name

0-yibai committed:
a skeletal version that greets the user and exits

0-yibai committed:
added echo and "bye"-exit

0-yibai committed:
allow adding tasks into list

0-yibai committed:
allow marking tasks as done or undone

0-yibai committed:
added functionality: parse todo, deadline and event

0-yibai committed:
updated message printed when adding and listing tasks

0-yibai committed:
add UI auto test cases

0-yibai committed:
handled some exceptions on inputs for mark and tasks

0-yibai committed:
allow users to delete tasks

0-yibai committed:
used enum to store actions

0-yibai committed:
use Printer class to handle all message printing

0-yibai committed:
file management: save and load tasks

0-yibai committed:
Merge branch 'save-tasks'

add a new feature

0-yibai committed:
store date and time as objects

0-yibai committed:
Abstract out ui and task list

0-yibai committed:
Add packages

0-yibai committed:
Merge branch 'use-package'

Using packages improves organization of code.

Class and enum files are grouped into 3 packages, namely
wis, wis.task and wis.util.

* wis.task contains all classes related to tasks and task
list
* wis.util contains the useful utility classes
* wis contains the backbone classes

Due to packaging, paths of files have been changed. Hence,
support for text-ui-test is suspended.

0-yibai committed:
Merge branch 'add-gradle-support'

0-yibai committed:
Modify paths for build.gradle

0-yibai committed:
Add junit test for InputParser.java

0-yibai committed:
Add junit test for Task.java

0-yibai committed:
Merge branch 'add-junit-test'

0-yibai committed:
Change path of tasks storage

RyanNgWH commented on docs/README.md:
```diff
@@ -1,29 +1,30 @@
 
-Description of the feature.
+// Product intro goes here
 
-### Feature-XYZ
+## Adding deadlines
 
-Description of the feature.
+// Describe the action and its outcome.
 
-## Usage
+// Give examples of usage
 
-### `Keyword` - Describe action
+Example: `keyword (optional arguments)`
 
-Describe the action and its outcome.
+// A description of the expected outcome goes here
```
Nice use of comments to note things to do in the future!

## Tip:
Adding the keyword `TODO:` in front of comments could be useful as some IDEs will list them in the `Problems` tool window. This will help you remember which parts of the code you have marked as todos.

```java
// TODO: Update the title above to match the actual product name

// TODO: Product screenshot goes here

// TODO: Product intro goes here
```

### Example:
<img width="1188" alt="image" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/44436551/45d92153-bfb3-4e37-b61b-31219f6416d2">

RyanNgWH commented on src/main/java/wis/Ui.java:
```diff
@@ -0,0 +1,144 @@
+            Printer.printActionAttach(Action.ADD_DEADLINE, deadline, tasks);
+        } catch (InputMismatchException e) {
+            WisException.ActionExceptionHandler(Action.ADD_DEADLINE);
+        } catch (DateTimeParseException e) {
+            WisException.DateTimeExceptionHandler(e);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            WisException.DateTimeExceptionHandler(e);
```
Could consider catching both exceptions in the same catch block since handling of the exception is similar

```java
        } catch (DateTimeParseException | ArrayIndexOutOfBoundsException  e) {
            WisException.DateTimeExceptionHandler(e);
        }
```

0-yibai replied:
This is the syntax I am looking for! I didn't know such syntax was allowed in Java.

RyanNgWH commented on src/main/java/wis/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Locale;
+
+public class Deadline extends Task {
+    LocalDateTime time;
```
Perhaps use a more descriptive variable name such as `dueDateTime` as `time` could be a little ambiguous

0-yibai replied:
Good suggestion. Noted.

RyanNgWH commented on src/main/java/wis/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Override
+    public String toString() {
+        // Format: Oct 15 2019, 1800
```
Great use of comments to illustrate the format!

RyanNgWH commented on src/main/java/wis/task/Deadline.java:
```diff
@@ -0,0 +1,33 @@
+    @Override
+    public String toString() {
+        // Format: Oct 15 2019, 1800
+        return String.format("[D]%s (by: %s %d %d, %s)",
+                super.toString(),
+                time.getMonth().getDisplayName(TextStyle.SHORT, Locale.ENGLISH).toUpperCase(),
+                time.getDayOfMonth(),
+                time.getYear(),
+                LocalTime.of(time.getHour(), time.getMinute()));
```
Nice use of line breaks to improve readability of long statement! 👍🏼

RyanNgWH commented on src/main/java/wis/task/TaskList.java:
```diff
@@ -0,0 +1,55 @@
+import wis.util.Printer;
+
+public class TaskList {
+    private ArrayList<Task> list;
```
Consider using `tasks` as the variable name instead to describe what this list contains

RyanNgWH commented on src/main/java/wis/util/InputParser.java:
```diff
@@ -0,0 +1,99 @@
+
+public class InputParser extends Parser {
+    public static Action parseAction(String input, String[] words) {
+        if (input.isEmpty()) {
+            return Action.NONE;
+        }
+        if (input.equals("list")) {
+            return Action.LIST;
+        }
+        if (input.equals("bye")) {
+            return Action.BYE;
+        }
+        if (words[0].equals("mark")) {
+            return Action.MARK;
+        }
+        if (words[0].equals("unmark")) {
+            return Action.UNMARK;
+        }
+        if (words[0].equals("delete")) {
+            return Action.DELETE;
+        }
+        if (words[0].equals("todo")) {
+            return Action.ADD_TODO;
+        }
+        if (words[0].equals("deadline")) {
+            return Action.ADD_DEADLINE;
+        }
+        if (words[0].equals("event")) {
+            return Action.ADD_EVENT;
+        }
+        return Action.INVALID;
```
Perhaps use a `switch...case` block to improve readability?

```java
switch (words[0]) {
case "list":
    return Action.LIST;

case "bye":
    return Action.BYE;
...
default:
    return Action.INVALID;
```

0-yibai replied:
Thank you for suggestion. Will change this.

RyanNgWH commented on src/main/java/wis/util/InputParser.java:
```diff
@@ -0,0 +1,99 @@
+        return parts.length == 2;
+    }
+
+    public static String[] parseDeadline(String input) {
+        String[] ret = new String[3];
+        Pattern pattern = Pattern.compile("^deadline ([^ ].*) /by ([^ ].*)$");
+        Matcher matcher = pattern.matcher(input);
+        if (matcher.find() && isSingleOccurrence("/by", input)) {
+            ret[0] = matcher.group(1);
+            ret[1] = matcher.group(2);
+        } else {
+            throw new InputMismatchException("Unknown input format");
+        }
+        return ret;
+    }
```
Maybe add a `throws InputMismatchException` in the method header to ensure that references handle the exception

```java
public static String[] parseDeadline(String input) throws InputMismatchException {
```

RyanNgWH commented on src/main/java/wis/util/Parser.java:
```diff
@@ -0,0 +1,4 @@
+package wis.util;
+
+public abstract class Parser {
+}
```
An empty abstract class seems redundant. What was your rationale for including this? Maybe add a comment or some javadocs to explain the purpose?

0-yibai replied:
I'm just trying to following this requirement 🥲
![Screenshot 2024-02-07 at 15 38 52](https://github.com/nus-cs2103-AY2324S2/ip/assets/110611841/20dd4ef3-0b21-4f87-9924-9811371719c9)

DingBao-sys replied:
I think its fine since in the future author can add methods to Parser class

RyanNgWH commented on src/main/java/wis/util/Printer.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public static void printActionAttach(Action action, Task task, TaskList tasks) {
+        switch (action) {
+            case ADD_TODO:
+            case ADD_DEADLINE:
+            case ADD_EVENT:
+                printDecorator();
+                Printer.println("Got it. I've added this task:");
+                Printer.printlnFurtherIndent(task.toString());
+                tasks.printTaskCount();
+                printDecorator();
+                break;
+            case DELETE:
+                printDecorator();
+                Printer.println("Noted. I've removed this task:");
+                Printer.printlnFurtherIndent(task.toString());
+                tasks.printTaskCount();
+                printDecorator();
+                break;
+            default:
+                throw new IllegalArgumentException("Illegal action argument provided.\n");
```
`case` statements should be inline with `switch` statement according to the coding guideline

Reference: https://se-education.org/guides/conventions/java/intermediate.html#layout
<img width="989" alt="image" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/44436551/6fd32b30-2e85-4788-8240-e56eaca33058">

0-yibai replied:
Noted. Thank you! (I forgot to change this)

RyanNgWH commented on src/main/java/wis/util/WisException.java:
```diff
@@ -0,0 +1,80 @@
+import wis.Action;
+
+public class WisException {
+    public static void ActionExceptionHandler(Action action) {
```
Would using a verb for the method name be a better idea? i.e `handleActionException` instead of `ActionExceptionHandler`

Should method names should be in camelCase instead of PascalCase?

Reference: https://nus-cs2103-ay2324s2.github.io/website/schedule/week4/topics.html#w4-6-code-quality-naming

<img width="739" alt="image" src="https://github.com/nus-cs2103-AY2324S2/ip/assets/44436551/36137fe3-4682-4ba8-8e69-4dc997188479">

0-yibai replied:
Noted. Thank you!

RyanNgWH commented on src/test/java/wis/task/TaskTest.java:
```diff
@@ -0,0 +1,23 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskTest {
+    private class TaskSlub extends Task {
```
Possibly misspelled `Stub` in `TaskSlub`?

0-yibai replied:
Yes, it's a misspell. Thank you for looking so carefully at my code...

RyanNgWH commented on src/test/java/wis/util/InputParserTest.java:
```diff
@@ -0,0 +1,111 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class InputParserTest {
```
Comprehensive test cases with comments to describe test details. Good work!

0-yibai replied:
Yeah it's quite tedious writing this Junit Test. Some online Regex testers could test this much faster. I feel that Junit is more suitable for testing algorithms, but there's not much algo in this project anyway...

RyanNgWH reviewed (commented):
Overall high quality code! 💯 

Just some minor coding standard violations. Adding explanations to some areas such as the empty abstract class would be great.

0-yibai committed:
Add JavaDoc

0-yibai committed:
Change method name and switch indent to standard

0-yibai committed:
Add search task by keyword functionality

0-yibai committed:
Merge branch 'branch-A-JavaDoc'

0-yibai committed:
Merge branch 'branch-A-CodingStandard' (conflict handled)

0-yibai committed:
Merge branch 'branch-A-Level-9'

DingBao-sys commented on src/main/java/wis/ChatBox.java:
```diff
@@ -0,0 +1,52 @@
+ * Main window running chat box of the application.
+ */
+public class ChatBox {
+    private Scanner scanner;
+    private Ui ui;
+    private TaskList tasks;
+    private boolean isExitSignal;
```
I like how you followed the naming conventions 👍

DingBao-sys commented on src/main/java/wis/Ui.java:
```diff
@@ -0,0 +1,167 @@
+    public boolean parseInput(TaskList tasks) {
+        String[] words = input.split(" ");
+        Action action = InputParser.parseAction(input, words);
+        switch (action) {
+        case NONE:
+            break;
+        case ADD_TODO:
+            addTodo(tasks);
+            break;
+        case ADD_DEADLINE:
+            addDeadline(tasks);
+            break;
+        case ADD_EVENT:
+            addEvent(tasks);
+            break;
+        case MARK:
+            mark(words, tasks);
+            break;
+        case UNMARK:
+            unmark(words, tasks);
+            break;
+        case DELETE:
+            delete(words, tasks);
+            break;
+        case LIST:
+            Printer.printActionAttach(Action.LIST, tasks);
+            break;
+        case FIND:
+            find(words, tasks);
+            break;
+        case BYE:
+            return true;
+        case INVALID:
+            WisException.actionExceptionHandler(Action.INVALID);
+            break;
+        default:
+            throw new IllegalArgumentException("Illegal action argument provided.\n");
+        }
```
Perhaps a bit of nitpicking to clear the indentation on switch blocks as per the conventions 😄

DingBao-sys commented on src/main/java/wis/Ui.java:
```diff
@@ -0,0 +1,167 @@
+            tasks.add(todo);
+            Printer.printActionAttach(Action.ADD_TODO, todo, tasks);
+        } catch (InputMismatchException e) {
+            WisException.actionExceptionHandler(Action.ADD_TODO);
+        }
+    }
+
+    private void addDeadline(TaskList tasks) {
+        try {
+            String[] parsedString = InputParser.parseDeadline(input);
+            Deadline deadline = new Deadline(parsedString[0],
+                    InputParser.parseDateTime(parsedString[1]));
+            tasks.add(deadline);
+            Printer.printActionAttach(Action.ADD_DEADLINE, deadline, tasks);
+        } catch (InputMismatchException e) {
+            WisException.actionExceptionHandler(Action.ADD_DEADLINE);
+        } catch (DateTimeParseException e) {
+            WisException.dateTimeExceptionHandler(e);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            WisException.dateTimeExceptionHandler(e);
+        }
+    }
+
+    private void addEvent(TaskList tasks) {
+        try {
+            String[] parsedString = InputParser.parseEvent(input);
+            Event event = new Event(parsedString[0],
+                    InputParser.parseDateTime(parsedString[1]),
+                    InputParser.parseDateTime(parsedString[2]));
+            tasks.add(event);
+            Printer.printActionAttach(Action.ADD_EVENT, event, tasks);
+        } catch (InputMismatchException e) {
+            WisException.actionExceptionHandler(Action.ADD_EVENT);
+        } catch (DateTimeParseException e) {
+            WisException.dateTimeExceptionHandler(e);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            WisException.dateTimeExceptionHandler(e);
+        }
+    }
+
```
I like how you seperated the creation of the different events into very clear functions. Although, may I suggest that you abstract away the construction of deadline event todo into their respective contractors so that you can have a simple addTask(TaskList tasks, Task task) method to do the adding. HAHAHA paiseh for class part 😃

DingBao-sys commented on src/main/java/wis/util/InputParser.java:
```diff
@@ -0,0 +1,136 @@
+    public static LocalDateTime parseDateTime(String string)
+            throws DateTimeParseException, ArrayIndexOutOfBoundsException {
+        // Input format: 2023-12-03 10:15
+        String[] strings = string.split(" ");
```
a bit of nitpick
perhaps
```suggestion
        String[] wordsInString = string.split(" ");
```

DingBao-sys commented on src/main/java/wis/Action.java:
```diff
@@ -0,0 +1,11 @@
+package wis;
+
+public enum Action {
+    NONE,
+    GREET, BYE,
+    LIST, FIND,
+    MARK, UNMARK,
+    ADD_TODO, ADD_DEADLINE, ADD_EVENT,
+    DELETE,
+    INVALID
+}
```
I like how you followed the convention of all caps

DingBao-sys reviewed (commented):
Overall I think really good job on the IP, in fact I learnt several amazing things from you like pattern hahah thanks man

0-yibai committed:
Correct switch indentation

0-yibai committed:
Turn CLI into GUI using Javafx

0-yibai committed:
Merge branch 'branch-Level-10'

0-yibai committed:
Allow save when exit from GUI

0-yibai committed:
bug fix: Remove keyword "bye" in GUI

0-yibai committed:
Add assertions to assert existence of a file

0-yibai committed:
refactor: Split Wis class into 2 classes

The original Wis class contains both code to render application and code
to get response.

The mix of distinct logic in Wis class causes weak SLAP.

Splitting the class allows us improve SLAP.

Let’s split Wis into 2 classes.
* WisApp which handles GUI initialization and rendering.
* Bridge which links the GUI to main Wis program.

0-yibai committed:
Improve code style as suggested by peer-review of PR

0-yibai committed:
refactor: Use guard statements and abstract out long conditions

If-else checks and long conditions causes low readability.

Lets,
* Turn if-else blocks into guard statements.
* Abstract out long conditional checks as separate methods.

Such Refactoring improves readability.

0-yibai committed:
refactor: Break down long method into smaller methods

0-yibai committed:
Merge pull request #1 from 0-yibai/branch-A-Assertions

Add assertions to assert existence of a file

0-yibai committed:
Merge branch 'master' into branch-A-CodeQuality

0-yibai committed:
Merge pull request #2 from 0-yibai/branch-A-CodeQuality

Merge branch improving code quality

0-yibai committed:
Add support for one-step undo

0-yibai committed:
Change dialog box to TextArea to allow user to copy

0-yibai committed:
Merge branch 'branch-C-Undo'

0-yibai committed:
Acknowledge code reuse in docs/README.md

0-yibai committed:
Update inline code reuse acknowledgement

0-yibai committed:
Display greet in GUI and add JavaDoc

0-yibai committed:
Add command descriptions in README

0-yibai committed:
Add images for app and documentation

0-yibai committed:
bug fix: Use remote path for markdown image

0-yibai committed:
Update remote path for markdown image

0-yibai committed:
Update remote path for ui screenshot image

0-yibai committed:
Change product screenshot file name

0-yibai committed:
Modify path for README

0-yibai committed:
Correct minor spelling mistake

0-yibai committed:
bug fix: Keep task count unchanged when deleting task
